source/Octopus.Tentacle.Tests.Integration/Util/SerilogLoggerBuilder.cs

```diff
@@ -2,48 +2,91 @@
 using System.Collections.Concurrent;
 using System.Diagnostics;
 using System.IO;
-using System.Text;
-using Halibut.Diagnostics;
+using System.Linq;
+using System.Security.Cryptography;
 using NUnit.Framework;
+using Octopus.Tentacle.Tests.Integration.Support;
+using Octopus.Tentacle.Tests.Integration.Support.ExtensionMethods;
 using Octopus.Tentacle.Util;
 using Serilog;
 using Serilog.Core;
+using Serilog.Events;
 using Serilog.Formatting.Display;
 
 namespace Octopus.Tentacle.Tests.Integration.Util
 {
     public class SerilogLoggerBuilder
     {
         public static readonly ConcurrentDictionary<string, Stopwatch> TestTimers = new();
+        static readonly ILogger Logger;
+        static readonly ConcurrentDictionary<string, TraceLogFileLogger> TraceLoggers = new();
+        static readonly ConcurrentBag<string> HasLoggedTestHash = new();
 
-        StringBuilder? stringBuilder;
+        TraceLogFileLogger? traceFileLogger;
 
-        public SerilogLoggerBuilder WithLoggingToStringBuilder(StringBuilder stringBuilder)
+        static SerilogLoggerBuilder()
         {
-            this.stringBuilder = stringBuilder;
+            const string teamCityOutputTemplate =
+                "{TestHash} "
+                + "{Timestamp:HH:mm:ss.fff zzz} "
+                + "[{ShortContext}] "
+                + "{Message}{NewLine}{Exception}";
+
+            const string localOutputTemplate =
+                "{Timestamp:HH:mm:ss.fff zzz} "
+                + "[{ShortContext}] "
+                + "{Message}{NewLine}{Exception}";
+
+            var nUnitOutputTemplate = TeamCityDetection.IsRunningInTeamCity()
+                ? teamCityOutputTemplate
+                : localOutputTemplate;
+
+            Logger = new LoggerConfiguration()
+                .MinimumLevel.Verbose()
+                .WriteTo.Sink(new NonProgressNUnitSink(new MessageTemplateTextFormatter(nUnitOutputTemplate)), LogEventLevel.Debug)
+                .WriteTo.Sink(new TraceLogsForFailedTestsSink(new MessageTemplateTextFormatter(localOutputTemplate)))
+                .CreateLogger();
+        }
+
+        public SerilogLoggerBuilder SetTraceLogFileLogger(TraceLogFileLogger logger)
+        {
+            this.traceFileLogger = logger;
             return this;
         }
         
         public ILogger Build()
         {
             // In teamcity we need to know what test the log is for, since we can find hung builds and only have a single file containing all log messages.
-            var testName = "";
-            if (TeamCityDetection.IsRunningInTeamCity())
+            var testName = TestContext.CurrentContext.Test.FullName;
+            var testHash = CurrentTestHash();
```

Comment:
Ported the TestHash with this change as well as it's a good change to reduce log line length